core/src/hir/methods.rs

```diff
@@ -36,6 +38,60 @@ pub struct Method {
     pub attrs: Attrs,
 }
 
+pub trait CallbackInstantiationFunctionality {
+    fn get_input_types(&self) -> Result<impl Iterator<Item = &Type<OutputOnly>>, TyPositionErr>; // the types of the parameters
+    fn get_output_type(&self) -> Result<&Option<Type>, TyPositionErr>;
+}
+
+#[derive(Debug)]
+#[non_exhaustive]
+// Note: we do not support borrowing across callbacks
+pub struct Callback {
+    pub param_self: Option<ParamSelf>, // for now it'll be none, but when we have callbacks as object methods it'll be relevant
+    pub params: Vec<CallbackParam>,
+    pub output: Box<Option<Type>>, // this will be used in Rust (note: can technically be a callback, or void)
+}
+
+#[derive(Debug, Clone)]
+pub enum TyPositionErr {
+    OnlyAllowedForInputTypes,
+    OnlyAllowedForOutputTypes,
```

Comment:
thought: I don't think this error is used. Perhaps we can just return `Result<_, ()>` here for now?

Reply:
good call